### PR TITLE
Fix num_class parameter. Now it takes the value set. It allows multiclass tasks.

### PR DESCRIPTION
**Pull request by ivallesp**
Hi,

I was trying to make a multiclass model using the R API and I found that it didn't detect when I indicated the correct number of classes using the parameter "num_class". I digged into the code and I found that the num_class parameter was harcoded so that I fixed it.

Hope it helps!

### CODE REVIEW

File: R-package/R/lgb.Booster.R [hunk 1]
```diff
@@ -49,7 +49,11 @@ Booster <- R6Class(
...
       class(handle)     <- "lgb.Booster.handle"
       private$handle    <- handle
-      private$num_class <- 1L
+      if ("num_class" %in% names(params)){
+	private$num_class <- params$num_class
+      } else{
+	private$num_class <- 1L
+      }
       private$num_class <-
         lgb.call("LGBM_BoosterGetNumClasses_R", ret = private$num_class, private$handle)
```
Comment (wxchan) [inline R-package/R/lgb.Booster.R]: doesn't num_class set here? the function's name is 'GetNumClasses'

Comment (ivallesp) [inline R-package/R/lgb.Booster.R]: Something does not work as expected. Why private$num_class <- 1L only changing it to the code I wrote solves the problem. I don't know what the C function GetNumClasses exactly does. What I am sure is that setting the parameter num_class to the number of classes in the call of the lgb.train does not work.

**Comment by Laurae2** [comment]
Do you have a reproducible example?

I compared:

* Microsoft/LightGBM (master branch)
* ivallesp/LightGBM (master branch)

I am getting exactly the same results on iris (I restart R for installation and after installation to clean the session), tested with two different settings of parameter input:

```r
library(lightgbm)

data(iris)
iris$Species <- as.numeric(as.factor(iris$Species))-1
dtrain <- lgb.Dataset(data=as.matrix(iris[c(1:40, 51:90, 101:140), 1:4]), label=iris$Species[c(1:40, 51:90, 101:140)])
dtest <- lgb.Dataset.create.valid(dtrain, data=as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]), label=iris$Species[c(41:50, 91:100, 141:150)])
valids <- list(test=dtest)
params <- list(objective="multiclass", metric="multi_error", num_class=3)
model <- lgb.train(list(), dtrain, 100, valids, min_data=1, learning_rate=1, early_stopping_rounds=10, objective="multiclass", metric="multi_error", num_class=3)
predict(model, as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]))
model <- lgb.train(params, dtrain, 100, valids, min_data=1, learning_rate=1, early_stopping_rounds=10)
predict(model, as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]))
```

Training works:

```r
[LightGBM] [Info] Number of data: 120, number of features: 4
[LightGBM] [Info] No further splits with positive gain, best gain: 0.000000, leaves: 2
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 3
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 4
[1]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 2
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 2
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 2
[2]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[3]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[4]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[5]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[6]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[7]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[8]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[9]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[10]:	test's multi_error:0 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf, leaves: 1
[LightGBM] [Info] Stopped training because there are no more leafs that meet the split requirements.
[11]:	test's multi_error:0 
```

Prediction results:

```r
# Differences:
 [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[74] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

# Microsoft master:
c(0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.825901296354802, 
0.0870493518225991, 0.0870493518225991, 0.825901296354802, 0.0870493518225991, 
0.0870493518225991, 0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.825901296354802, 
0.0870493518225991, 0.0870493518225991, 0.825901296354802, 0.0870493518225991, 
0.0870493518225991, 0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.100551144301225, 0.73876858293458, 
0.160680272764195, 0.106983998915497, 0.786032002169005, 0.106983998915497, 
0.106983998915497, 0.786032002169005, 0.106983998915497, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.106983998915497, 0.786032002169005, 
0.106983998915497, 0.106983998915497, 0.786032002169005, 0.106983998915497, 
0.106983998915497, 0.786032002169005, 0.106983998915497, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.106983998915497, 0.786032002169005, 
0.106983998915497, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593, 0.0910962699146787, 
0.0969714746877288, 0.811932255397593, 0.0910962699146787, 0.0969714746877288, 
0.811932255397593, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593, 0.0910962699146787, 
0.0969714746877288, 0.811932255397593, 0.0910962699146787, 0.0969714746877288, 
0.811932255397593, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593)

# Your branch:
c(0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.825901296354802, 
0.0870493518225991, 0.0870493518225991, 0.825901296354802, 0.0870493518225991, 
0.0870493518225991, 0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.825901296354802, 
0.0870493518225991, 0.0870493518225991, 0.825901296354802, 0.0870493518225991, 
0.0870493518225991, 0.825901296354802, 0.0870493518225991, 0.0870493518225991, 
0.825901296354802, 0.0870493518225991, 0.0870493518225991, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.100551144301225, 0.73876858293458, 
0.160680272764195, 0.106983998915497, 0.786032002169005, 0.106983998915497, 
0.106983998915497, 0.786032002169005, 0.106983998915497, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.106983998915497, 0.786032002169005, 
0.106983998915497, 0.106983998915497, 0.786032002169005, 0.106983998915497, 
0.106983998915497, 0.786032002169005, 0.106983998915497, 0.106983998915497, 
0.786032002169005, 0.106983998915497, 0.106983998915497, 0.786032002169005, 
0.106983998915497, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593, 0.0910962699146787, 
0.0969714746877288, 0.811932255397593, 0.0910962699146787, 0.0969714746877288, 
0.811932255397593, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593, 0.0910962699146787, 
0.0969714746877288, 0.811932255397593, 0.0910962699146787, 0.0969714746877288, 
0.811932255397593, 0.0910962699146787, 0.0969714746877288, 0.811932255397593, 
0.0910962699146787, 0.0969714746877288, 0.811932255397593)
```

File: R-package/R/lgb.Booster.R (hunk 1 above)
Comment (wxchan) [inline R-package/R/lgb.Booster.R]: I am not very familiar with R-package details, but the logic should be same as python-package. All the parameters including num_class is set at [L26](https://github.com/Microsoft/LightGBM/pull/289/files/ed379175b0c5f34ff23d4313ffa9a55df1b878da#diff-acaf2c3176205deafe7683a35b929405R26) to C++. And R-package gets num_class from C++ here.

Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: Workflow is like this:

* Get param list
* Apply param list to booster
* (...)
* num_class is set set to default value (if it doesn't exist, then it will be 1 at the end because it is the default value)
* num_class is harvested from booster else it gives the default value

It should be working already that way, can you post a reproducible example?

**Comment by ivallesp** [comment]
Thank you very much, @Laurae2. I will try as soon as I have a moment your example to see if it works in my current installation!.

**Comment by ivallesp** [comment]
Hi,

I reinstalled the current version and I followed the example of @Laurae2. My conclusions:

- It works fine! Same predictions
- I reproduced my issue. If you don't specify the num_class parameter in your first call to the lgb.train function, you notice and add it and then you re-launch the code, it keeps giving you the error:

> Error in lgb.call("LGBM_BoosterCalcNumPredict_R", ret = npred, private$handle,  : 
  api error: Number of classes should be specified and greater than 2 for multiclass training

Some code:

`library(lightgbm)

data(iris)
iris$Species <- as.numeric(as.factor(iris$Species))-1
dtrain <- lgb.Dataset(data=as.matrix(iris[c(1:40, 51:90, 101:140), 1:4]), label=iris$Species[c(1:40, 51:90, 101:140)])
dtest <- lgb.Dataset.create.valid(dtrain, data=as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]), label=iris$Species[c(41:50, 91:100, 141:150)])
valids <- list(test=dtest)
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass") # Error, remove after launching the first time
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass", num_class=3)
predict(model, as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]))
`

So if you launch it, remove the line indicated and relaunch it without restarting R, it keeps giving you the error.

Actually my code didn't resolve the error, I was too impulsive :-D.

**Comment by wxchan** [comment]
I think the error message is clear: Number of classes should be **`specified`**.

**Comment by Laurae2** [comment]
@wxchan He meant if you re-run the code without cleaning the environment, it will still error.

I suspect pointers not being freed up in time, or similar. I encountered this issue several times in LightGBM's R package, where if you did an error in setting up your training, you will always get an error until a strange message appears on console (something being freed up) or you restart the R (which is what @ivallesp describes).

```r
# This CANNOT work
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass")

# This MUST work, but we will see it doesn't if the one that cannot work is run
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass", num_class=3)
```

@ivallesp See the workaround below for temporary solution. I found no other reliable method.

Running this code will error out no matter what you do (unless the environment decides to fully clean itself, which requires freeing some pointers which are not freed out using garbage collection?), unless you set my `FAILURE_CHECK` to `FALSE` (which destroys lightgbm environment in the If-Then-Else in the middle):

```r
FAILURE_CHECK <- TRUE # set this to FALSE and you will see it will not error out

library(lightgbm)

# Load data
data(iris)
iris$Species <- as.numeric(as.factor(iris$Species))-1
dtrain <- lgb.Dataset(data=as.matrix(iris[c(1:40, 51:90, 101:140), 1:4]), label=iris$Species[c(1:40, 51:90, 101:140)])
dtest <- lgb.Dataset.create.valid(dtrain, data=as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]), label=iris$Species[c(41:50, 91:100, 141:150)])
valids <- list(test=dtest)

# ERROR!
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass")

# ERROR but it is not supposed to throw error!
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass", num_class=3)

# Clean up environment aggressively
rm(list = (ls())[!names(ls()) %in% "FAILURE_CHECK"])

# FAILURE_CHECK: if TRUE it will error out, if FALSE it will not error out later
if (FAILURE_CHECK == TRUE) {
  gc(verbose = TRUE) # get rid of all memory used previously, still errors out later
} else {
  detach("package:lightgbm", unload=TRUE) # destroys LightGBM environment, which also destroys also pointers
  gc(verbose = TRUE) # get rid of all memory used previously
  library(lightgbm) # reload package
}

# Load data again
data(iris)
iris$Species <- as.numeric(as.factor(iris$Species))-1
dtrain <- lgb.Dataset(data=as.matrix(iris[c(1:40, 51:90, 101:140), 1:4]), label=iris$Species[c(1:40, 51:90, 101:140)])
dtest <- lgb.Dataset.create.valid(dtrain, data=as.matrix(iris[c(41:50, 91:100, 141:150), 1:4]), label=iris$Species[c(41:50, 91:100, 141:150)])
valids <- list(test=dtest)

# ERROR if FAILURE_CHECK = TRUE, but it is not supposed to throw error!
model <- lgb.train(list(), dtrain, nrounds=100, objective="multiclass", num_class=3)
```

**Comment by wxchan** [comment]
@Laurae2 I realized it, ignore me.

**Comment by ivallesp** [comment]
Amazing, @Laurae2! Thank you very much; very appreciated.

**Comment by guolinke** [comment]
@Laurae2 So this PR is not need? Right?

**Comment by Laurae2** [comment]
@guolinke To close, yes.
